File: Backend/lib/src/testFixtures/kotlin/com/github/bratek20/ostium/gamesetup/tests/ImplTest.kt
package com.github.bratek20.ostium.gamesetup.tests

import com.github.bratek20.architecture.context.someContextBuilder
import com.github.bratek20.logs.LoggerMock
import com.github.bratek20.logs.LogsMocks
import com.github.bratek20.ostium.gamecomponents.fixtures.creatureCardId
import com.github.bratek20.ostium.gamesetup.api.GameSetupApi
import com.github.bratek20.ostium.gamesetup.api.RowType
import com.github.bratek20.ostium.gamesetup.context.GameSetupImpl
import com.github.bratek20.ostium.gamesetup.fixtures.assertGame
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test

open class GameSetupImplTest {
    class Context(
        val api: GameSetupApi,
        val loggerMock: LoggerMock
    )
    open fun createContext(): Context {
        val c = someContextBuilder()
            .withModules(
                GameSetupImpl(),
                LogsMocks()
            ).build()

        return Context(
            api = c.get(GameSetupApi::class.java),
            loggerMock = c.get(LoggerMock::class.java)
        )
    }

    private lateinit var api: GameSetupApi
    private lateinit var loggerMock: LoggerMock

    @BeforeEach
    fun setup() {
        val c = createContext()
        this.api = c.api
        this.loggerMock = c.loggerMock
    }

    @Test
    fun `start game`() {
        val game = api.startGame()

        loggerMock.assertInfos(
            "Game started"
        )
        assertGame(game) {
            hand = {
                cards = listOf(
                    {
                        id = "Mouse1"
                    },
                    {
                        id = "Mouse2"
                    }
                )
            }
            table = {
                gateDurabilityCard = {
                    myMarker = 15
                    opponentMarker = 15
                }
                attackRowEmpty = true
                defenseRowEmpty = true
                gateCard = {
                    destroyed = false
                }
            }
        }
    }

    @Test
    fun `play card`() {
        api.startGame()

        val game = api.playCard(creatureCardId("Mouse1"), RowType.ATTACK)

        assertGame(game) {
            hand = {
                cards = listOf {
                    id = "Mouse2"
                }
            }
            table = {
                attackRow = {
                    id = "Mouse1"
                }
            }
        }
    }

    @Test
    fun `move card`() {
        api.startGame()
        api.playCard(creatureCardId("Mouse1"), RowType.ATTACK)

        val game = api.moveCard(creatureCardId("Mouse1"), RowType.ATTACK, RowType.DEFENSE)

        assertGame(game) {
            hand = {
                cards = listOf {
                    id = "Mouse2"
                }
            }
            table = {
                attackRowEmpty = true
                defenseRow = {
                    id = "Mouse1"
                }
            }
        }
    }
}

File: Frontend/Logic/OstiumBackend.Tests/GameSetup/Tests/ImplTest.cs
using System;
using System.Collections.Generic;
using B20.Ext;
using B20.Infrastructure.HttpClient.Integrations;
using GameComponents.Api;
using GameSetup;
using GameSetup.Api;
using GameSetup.Impl;
using GameSetup.Web;
using HttpClientModule.Api;
using HttpClientModule.Impl;
using Xunit;

namespace OstiumBackend.Tests.GameSetup.Tests
{
    public class GameSetupImplTest
    {
        private GameSetupApi CreateApi()
        {
            // Assuming someContextBuilder() is a placeholder for the actual implementation
            // that provides necessary context and module setup for GameSetupApi.
            return new GameSetupApiLogic();
        }

        [Fact]
        public void StartGame_ShouldInitializeCorrectly()
        {
            var api = CreateApi();
            var game = api.StartGame();

            Asserts.AssertGame(game, expected =>
            {
                expected.Hand = hand =>
                {
                    hand.Cards = new List<Action<GameComponents.Diffs.ExpectedCreatureCard>>
                    {
                        card => { card.Id = "Mouse1"; },
                        card => { card.Id = "Mouse2"; }
                    };
                };

                expected.Table = table =>
                {
                    table.GateDurabilityCard = card =>
                    {
                        card.MyMarker = 15;
                        card.OpponentMarker = 15;
                    };
                    table.AttackRowEmpty = true;
                    table.DefenseRowEmpty = true;
                    table.GateCard = gate =>
                    {
                        gate.Destroyed = false;
                    };
                };
            });
        }
        


        [Fact]
        public void PlayCard_ShouldPlaceCardInAttackRow()
        {
            var api = CreateApi();
            api.StartGame();

            var game = api.PlayCard(new CreatureCardId("Mouse1"), RowType.ATTACK);

            Asserts.AssertGame(game, expected =>
            {
                expected.Hand = hand =>
                {
                    hand.Cards = new List<Action<GameComponents.Diffs.ExpectedCreatureCard>>
                    {
                        card => { card.Id = "Mouse2"; }
                    };
                };

                expected.Table = table =>
                {
                    table.AttackRow = row =>
                    {
                        row.Id = "Mouse1";
                    };
                };
            });
        }

        [Fact]
        public void MoveCard_ShouldMoveCardToDefenseRow()
        {
            var api = CreateApi();
            api.StartGame();
            api.PlayCard(new CreatureCardId("Mouse1"), RowType.ATTACK);

            var game = api.MoveCard(new CreatureCardId("Mouse1"), RowType.ATTACK, RowType.DEFENSE);

            Asserts.AssertGame(game, expected =>
            {
                expected.Hand = hand =>
                {
                    hand.Cards = new List<Action<GameComponents.Diffs.ExpectedCreatureCard>>
                    {
                        card => { card.Id = "Mouse2"; }
                    };
                };

                expected.Table = table =>
                {
                    table.AttackRowEmpty = true;
                    table.DefenseRow = row =>
                    {
                        row.Id = "Mouse1";
                    };
                };
            });
        }
        
        [Fact]
        public void MoveCard_ShouldSwapCards()
        {
            var api = CreateApi();
            api.StartGame();
            api.PlayCard(new CreatureCardId("Mouse1"), RowType.ATTACK);
            api.PlayCard(new CreatureCardId("Mouse2"), RowType.DEFENSE);

            var game = api.MoveCard(new CreatureCardId("Mouse1"), RowType.ATTACK, RowType.DEFENSE);

            Asserts.AssertGame(game, expected =>
            {
                expected.Table = table =>
                {
                    table.AttackRow = row =>
                    {
                        row.Id = "Mouse2";
                    };
                    table.DefenseRow = row =>
                    {
                        row.Id = "Mouse1";
                    };
                };
            });
        }
    }
}
